### PR TITLE
Save correctly selective masking parameters

### LINKED ISSUE #33277: Symbol masking for labels seems quite broken
Possibly I'm totally misusing this feature, but I can't get it to actually DO anything!

The settings are lost just by closing dialogs, and regardless of which combination of checkboxes I tick there's no visible impact on the map...

Peek 2019-12-07 19-59

### PR DESCRIPTION
**Pull request by troopa81**
## Description

This PR fixes #33277. It fixes several things:

- It applies masking widget settings from the vector properties

- The edit mask settings were not serialized in the project and so need to be unchecked every time we display the widget. This checkbox was used to defer widget populate which takes time #35557. I propose to delay the populate when the widget is first shown. 

- If user selects a masked symbol layer but no mask source (or vice versa) the configuration is not saved because data model associates sources with layers. I propose to display a warning message when the configuration is not consistent.

msgSelectiveMasking

### CODE REVIEW

File: src/gui/qgsmaskingwidget.cpp [hunk 1]
```diff
@@ -28,29 +28,54 @@
...
-  connect( mMaskSourcesWidget, &QgsMaskSourceSelectionWidget::changed, this, [&]()
+    mMessageBar->popWidget( mMessageBarItem );
+    delete mMessageBarItem;
+    mMessageBarItem = nullptr;
+  }
+  else if ( !mMessageBarItem && printMessage )
   {
-    emit widgetChanged();
-  } );
+    mMessageBarItem = new QgsMessageBarItem( tr( "You need to select both mask sources and masked symbol layers. If not, all configuration will be lost" ), Qgis::Warning, 0, this );
```
Comment (mhugo) [inline src/gui/qgsmaskingwidget.cpp:54]: "If not, all configuration will be lost" => ", otherwise the configuration will be lost" ?

**Comment by mhugo** [comment]
Loooks good to me, thanks !

File: src/gui/qgsmaskingwidget.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/gui/qgsmaskingwidget.cpp:54]: This message is way too long for the message bar anyway. It needs to be made much more concise

File: src/gui/qgsmaskingwidget.h [hunk 2]
```diff
@@ -47,10 +49,21 @@ class GUI_EXPORT QgsMaskingWidget: public QgsPanelWidget, private Ui::QgsMasking
...
+    void onSelectionChanged();
+
   private:
     QgsVectorLayer *mLayer = nullptr;
     //! Populate the mask source and target widgets
     void populate();
+
+    QgsMessageBarItem *mMessageBarItem = nullptr;
+    bool _mustPopulate = false;
 };
```
Comment (nyalldawson) [inline src/gui/qgsmaskingwidget.h:67]: mMustPopulate

File: src/gui/qgsmaskingwidget.cpp (hunk 1 above)
Comment (troopa81) [reply src/gui/qgsmaskingwidget.cpp]: I changed it to "Select both sources and symbol layers or your configuration will be lost"

File: src/gui/qgsmaskingwidget.h [hunk 3]
```diff
@@ -47,10 +49,24 @@ class GUI_EXPORT QgsMaskingWidget: public QgsPanelWidget, private Ui::QgsMasking
     //! Emitted when a change is performed
     void widgetChanged();
 
+  protected:
+
+    void showEvent( QShowEvent * ) override;
+
+  protected slots:
```
Comment (m-kuhn) [inline src/gui/qgsmaskingwidget.h:56]:
```suggestion
  private slots:
```

File: src/gui/qgsmaskingwidget.cpp [hunk 4]
```diff
@@ -28,29 +28,54 @@
...
   QgsPanelWidget( parent )
 {
   setupUi( this );
+}
+
+void QgsMaskingWidget::onSelectionChanged()
+{
+  // display message if configuration is not consistent
+  bool printMessage =
+    ( mMaskTargetsWidget->selection().empty() && !mMaskSourcesWidget->selection().empty() )
```
Comment (m-kuhn) [inline src/gui/qgsmaskingwidget.cpp:43]: Or simply `mMaskTargetsWidget->selection().empty() != mMaskSourcesWidget->selection().empty()`?

File: src/gui/qgsmaskingwidget.cpp [hunk 5]
```diff
@@ -28,29 +28,54 @@
...
 
-  connect( mMaskTargetsWidget, &QgsSymbolLayerSelectionWidget::changed, this, [&]()
+  if ( mMessageBarItem && !printMessage )
   {
-    mMaskSourcesWidget->setEnabled( ! mMaskTargetsWidget->selection().isEmpty() );
-    emit widgetChanged();
-  } );
-  connect( mMaskSourcesWidget, &QgsMaskSourceSelectionWidget::changed, this, [&]()
+    mMessageBar->popWidget( mMessageBarItem );
+    delete mMessageBarItem;
```
Comment (m-kuhn) [inline src/gui/qgsmaskingwidget.cpp:49]: What happens if the user clicks the `X` on the item? is there a risk of a crash?

Comment (troopa81) [reply src/gui/qgsmaskingwidget.cpp:47]: yes. I added a QPointer

**Pull request merged**